### DISCUSSION TITLE
Golang tutorial or framework easy setup

### DISCUSSION THREAD

**Discussion by The-Code-Monkey** [vercel/vercel#6562] [Ideas]
A tutorial or framework setup for a golang server would be good as of right now I am unsure how to set one up

**Comment by barelyhuman** [comment]
Not sure, though there are already examples in the repo for all languages. 

Here's mine that I use for small microservices, if that helps. 

https://github.com/barelyhuman/next-go-vercel

**Reply by The-Code-Monkey** [reply]
What commands are you running on vercel to start go as it keeps telling me that go isn't installed, what are your start and build commands for the go server within vercel?

**Reply by barelyhuman** [reply]
To clarify, vercel does serveless functions based on 2 folders, the `api` folder and if you are using next.js , then the `pages/api` folder. 

To initialize go lang, you just need to create `.go` files that act as microservices (pattern in the example repo) , for vercel to actually setup the runtime, in case you are need a custom go binary to be run everytime something is built, you need a `go.mod` file in the project. 

The below is a complex scenario but to put in simple words. 

It compiles the go files then runs it during build time (this is done by adding the `go run .`) in the vercel dashboard under 
Projects> [Your Project] > Settings > Build Command > `amazon-linux-extras install golang1.11; go run .`

This is how an older state of my website used go to build the website on each deploy. 

https://github.com/barelyhuman/reaper.im/tree/5476959a0f459029c790235487f6b1774adc0268

### Server 

You don't need to create your own go server or handlers using `net/http` , that is taken care of by the vercel bridges and you just need to provide the bridges with the needed functions. 

Eg: 
if I need https://somedomain.com/api/ping  , I create a file named `ping.go` in a folder named `api` and `ping.go` has the needed handling. 

You can read more about it here 
https://vercel.com/docs/runtimes#official-runtimes/go

**Reply by The-Code-Monkey** [reply]
Hmm not sure how to set mine up due to having a more separate API system built on graphql,. Here is my repo if you could help out, it's my first time using go, https://github.com/The-Code-Monkey/Website-Core-Ecommerce API stuff is in the backend folder.

The dB currently is running locally within docker using mysql

**Reply by barelyhuman** [reply]
Went through the repo, as I mentioned , you have a traditional backend which would need a proper server, something like heroku, if you wish to use the graphql on vercel functions, you'll have to make quite a few modifications to the code. 

There's a good writeup from hasura regarding this which you can take references from. 
https://hasura.io/learn/graphql/nextjs-fullstack-serverless/introduction/
It's focused on Next.js but you can pick the general idea of graphql serverless 

or this, 
https://www.serverless.com/blog/running-scalable-reliable-graphql-endpoint-with-serverless

Hope that helps you out :)
